incidents/2018-04-24-journal-outage.md

```diff
@@ -0,0 +1,43 @@
+# 2018-04-24 9.5hr journal outage
+
+## Description
+
+All nodes on journal--prod failed highstate, journal left broken for many hours.
+
+## Timeline
+
+(UTC+9:30)
+
+6:54am - First alert from Loggly that highstate had failed
+
+9:20am - Investigation into highstate on journal--prod
+
+9:27am - Website check, looks like it's running
+
+10:35am - Finished investigation, email to techalerts. Conclusion: low-priority bug.
+
+3:54pm - Alert from Chris that journal is actually down
+
+4:01pm - Chris merges fix
+
+4:03pm - Updating master-server and journal--prod nodes
+
+4:08pm - Journal responding to requests again
+
+4:15pm - All nodes updated
+
+## Contributing Factors
+
+* CDN fooled me into thinking website was running
+* it's common for failed states to allow the application to continue working
+* changes to application passed CI
+* apart from highstate failure reported by Loggly, no other (related) alerts to journal being down
+
+## Impact
+
+Journal was (mostly) down from approx 7am to 4pm (9hrs)
+
+## Corrective actions
+
+* builder should halt serial operations on error. We would have had 2 of 3 nodes working.
```

Comment:
that's a good point. hmm. I was literally looking at the concurrency code in lifecycle when Chris came online, probably why I jumped to that conclusion.

was there talk of having Alfred manage the updates?

Reply:
Yes https://elifesciences.atlassian.net/browse/ELPP-3213 which was postponed due to too many instances and a failure rate high enough that something would fail in the {{ci}} or {{end2end}} phases every day. Will have to do more work on containers first to relieve us from the high number of nodes.

Reply:
we could have the master server run the daily highstate on it's minions?

Reply:
the problem is not in triggering the highstate in parallel on many instances (it's a single command), but rather in getting all of them to be green; there seem to be always some timeout, or APT issue, or Pip package, etc. that fails at least one of them. My strategy is to reduce huge formulas moving some responsibilities to immutable containers so that we have to update fewer EC2 instances and each of them has fewer moving parts that can fail. Unfortunately it's a slow process...

Reply:
I'm here to help if there is some goal we're working towards. From my perspective on the other side of the world, the containerisation effort seems quite adhoc and disorganised. Is there an epic-level ticket you can point me to?

For the time being we can leave this problem as-is or we can figure out how to prevent a highstate from happening if the previous node isn't up.

```
if node == 1: do highstate
else: ping (self.node - 1) and do highstate
```

that logic could exist in the daily highstate script which itself could be run at an offset if it's node is > 1. the ping is some script that knows how to check itself and report healthiness (which I think we have a ticket for).

Reply:
https://elifesciences.atlassian.net/browse/ELPP-3314 is the place

I don't see a clean way to have nodes communicate while maintaining their agnosticism about the rest of the cluster; one thing that we do now is to run daily highstate at different time on even/odd numbered nodes so that they don't all go out at the same time. We could have a larger time differential like 12 hours:
https://github.com/elifesciences/builder-base-formula/blob/master/elife/daily-system-updates.sls#L30-L36

Reply:
that could work if NewRelic picks up an app is down after a bad highstate.

worst case scenario is a breaking change introduced by the UK on a Friday causing outage until Sunday (my Monday).

Reply:
from retrospective: try to run daily updates pipeline on user-facing projects (journal, maybe journal-cms) rather than crontabs. This limits the fallout to a few nodes rather than the whole ~100 instances.